### ISSUE TITLE
Weird inconsistent behaviour with dynamics value

### ISSUE THREAD

**Issue by VincentLanglet** [cssinjs/jss#1057]
__Codesandbox link:__ 
https://codesandbox.io/s/p2wljkm9zm

In this example you can see
```tsx
const styles = () =>
  createStyles({
    text: {
      backgroundColor: "red", // WORKING
      color: () => "blue", // WORKING
      "&:hover": {
        marginLeft: 20, // WORKING,
        marginRight: () => 20 // WORKING,
      }
    },
    someText: () => ({
      backgroundColor: "red", // WORKING
      color: () => "blue", // NOT WORKING
      "&:hover": {
        marginLeft: 20, // WORKING, 
        marginRight: () => 20 // WORKING,
      }
    }),
    otherText: {
      backgroundColor: "red", // WORKING
      color: () => "blue", // WORKING
      "&:hover": () => ({
        marginLeft: 20, // NOT WORKING,
        marginRight: () => 20 // NOT WORKING,
      })
    },
    finalText: () => ({
      backgroundColor: "red", // WORKING
      color: () => "blue", // NOT WORKING
      "&:hover": () => ({
        marginLeft: 20, // WORKING,
        marginRight: () => 20 // NOT WORKING,
      })
    })
  });
```

I understand the fact function inside function does not work, like color of someText, but in this case why marginRight of someText does work ?
And why marginLeft of otherText does not work ? Even weirder, it works for finalText...

If it works when it shouldn't at least it's not a big deal. But I think the case of marginLeft of otherText should be a correct one, and actually it doesn't work.

**Comment by HenriBeck** [comment]
`marginRight ` works because `&:hover` will become a separate rule on the sheet.
```js
someText: () => ({
      backgroundColor: "red", // WORKING
      color: () => "blue", // NOT WORKING
      "&:hover": {
        marginLeft: 20, // WORKING, 
        marginRight: () => 20 // WORKING,
      }
    }),
```
will become:
```js
someText: () => ({
      backgroundColor: "red", // WORKING
      color: () => "blue", // NOT WORKING
    }),
 "someTextSelector:hover": {
        marginLeft: 20, // WORKING, 
        marginRight: () => 20 // WORKING,
      }
```

**Comment by kof** [comment]
We should warn when user tries to do unsupported stuff when possible.

**Comment by HenriBeck** [comment]
I think why `marginLeft` inside `otherText` doesn't work is because we split dynamic rules from the static rules. I think if we would to do this just with jss it would work.

**Comment by kof** [comment]
Still we don't want people to use function values inside of function rules and I think we can warn about it.

**Comment by oliviertassinari** [comment]
If we can improve performance by not supporting prop render nesting, I'm all 💯 for it. Is styled components supporting it?

**Comment by kof** [comment]
We are not supporting value functions inside of rule functions, it just worked by accident I think.

**Comment by HenriBeck** [comment]
They are only working because of the way we are treating nested rules. I don't think there is a use case for supporting function values inside a function declaration.

**Comment by flucivja** [comment]
Hi, what in case there is needed to apply function value for media query?

Something like this:
```
tableColumn: {
  maxWidth: ({columnsLength}) => colMaxWidth(theme.screen.screenXs, columnsLength),
        
  [`@media screen and (min-width: ${theme.screen.screenSm})`]: {
    maxWidth: ({columnsLength}) => colMaxWidth(theme.screen.screenSm, columnsLength)
  }
}
```
The max-width outside of media query is properly calculated but inside of media query it is not.

btw putting media query outside of tableColumn does not help.
```
tableColumn: {
  maxWidth: ({columnsLength}) => colMaxWidth(theme.screen.screenXs, columnsLength)
},
[`@media screen and (min-width: ${theme.screen.screenSm})`]: {
  tableColumn: {
    maxWidth: ({columnsLength}) => colMaxWidth(theme.screen.screenSm, columnsLength)
  }
}
```

**Comment by kof** [comment]
@flucivja I think you misunderstood this issue and talking about another problem which should have been fixed in v10

**Comment by flucivja** [comment]
Ah sorry, my bad. Btw. is there issue for such problem already?

**Comment by kof** [comment]
I stated that the problem might have been fixed in v10, did you verify that?

**Comment by kof** [comment]
released in 10.1.0
